### PR TITLE
PERF: Switch `plugin:install_all_official` to clone plugins concurrently

### PR DESCRIPTION
**Pull request by tgxworld**
Why this change?

`plugin:install_all_official` is quite slow at the moment taking roughly
1 minute and 51 seconds on my machine. Since most of the time is spent
waiting on the network, we can actually speed up the Rake task
significantly by executing the cloning concurrently. With a 8 cores
machine, cloning all plugins will only take 15 seconds.

What does this change do?

This change wraps the `git clone` operation in the
`plugin:install_all_official` Rake task in a `Concurrent::Promise` which
basically runs the `git clone` operation in a Thread. The `--quiet`
option has also been added to `git clone` since running stuff
concurrently messes up the output. That could be fixed but it has been
determined to be not worth it since the output from `git clone` is
meaningless to us.

### CODE REVIEW

**Comment by tgxworld** [comment]
Screenshot from 2023-11-22 18-27-49

Interestingly the git clone is super fast on Github probably because the runners are super close to the source.

File: lib/tasks/plugin.rake [hunk 1]
```diff
@@ -24,16 +28,26 @@ task "plugin:install_all_official" do
...
+          abort
+        end
+
+        STDOUT.puts "Failed to clone #{repo}... trying again..."
+        retry
+      end
     end
   end
+
+  Concurrent::Promise.zip(*promises).value!
```
Comment (CvX) [inline lib/tasks/plugin.rake:49]: Is there a limit to the concurrency? Or will it try to completely saturate the bandwidth and the number of connections to a single host?

File: lib/tasks/plugin.rake [hunk 2]
```diff
@@ -24,16 +28,26 @@ task "plugin:install_all_official" do
...
+        system("git clone --quiet #{repo} #{path}", exception: true)
+      rescue StandardError
+        if attempts >= 3
+          failures << repo
+          abort
+        end
+
+        STDOUT.puts "Failed to clone #{repo}... trying again..."
+        retry
+      end
```
Comment (CvX) [inline lib/tasks/plugin.rake:45]:
```suggestion
    Concurrent::Promise.execute do
      attempts ||= 1
      STDOUT.puts("Cloning '#{repo}' to '#{path}'...")
      system("git clone --quiet #{repo} #{path}", exception: true)
    rescue StandardError
      if attempts == 3
        failures << repo
        abort
      end

      STDOUT.puts "Failed to clone #{repo}... trying again..."
      attempts += 1
      retry
    end
```

🏌️

**Review by CvX — APPROVED** [review]

File: lib/tasks/plugin.rake (hunk 1 above)
Comment (tgxworld) [reply lib/tasks/plugin.rake:49]: It will spin X number of threads where X is equal to the number of CPU cores on the host. We could set an upper limit here but I doubt anyone will be running this one a CPU with enough cores to cause problems.

Comment (CvX) [reply lib/tasks/plugin.rake:49]: Ah, nice!

Comment (davidtaylorhq) [reply lib/tasks/plugin.rake:49]: BTW, our internal CI host has 32 cores. Maybe still fine, but worth bearing in mind 👀

Comment (tgxworld) [reply lib/tasks/plugin.rake:49]: O I have 32 cores locally too and nothing broke. I’ll be alittle aggressive for now with the number of concurrent git clones but will scale it back down should we run into problems.

**Comment by tgxworld** [comment]
Thank you for the review @CvX and @davidtaylorhq

**Comment by tgxworld** [comment]
Small win on Github actions, we went from 22 seconds to clone all official plugins to 4 seconds.

Screenshot from 2023-11-22 21-44-36
